syntax = "proto3";

package raft;

service RaftService {
    rpc RequestVote (VoteArgs) returns (VoteReply) {}
    rpc AppendEntries (AppendArgs) returns (AppendReply) {}
}

message VoteArgs {
    int32 term = 1;
    int32 candidate_id = 2;
    int32 last_log_index = 3;
    int32 last_log_term = 4;
}

message VoteReply {
    int32 term = 1;
    bool vote_granted = 2;
}

message AppendArgs {
    int32 term = 1;
    int32 leader_id = 2;
    int32 prev_log_index = 3;
    int32 prev_log_term = 4;
    repeated LogEntry entries = 5;
    int32 leader_commit = 6;
}

message AppendReply {
    int32 term = 1;
    bool success = 2;
}

message LogEntry {
    int32 term = 1;
    string command = 2; 
    bytes data = 3;
}